zkevm-circuits/src/evm_circuit/execution/logs.rs

```diff
@@ -205,28 +203,27 @@ impl<F: Field> ExecutionGadget<F> for LogGadget<F> {
         self.memory_expansion
             .assign(region, offset, step.memory_word_size(), [memory_address])?;
 
-        let opcode = step.opcode.unwrap();
+        let opcode = step.opcode().unwrap();
         let topic_count = opcode.postfix().expect("opcode with postfix") as usize;
         assert!(topic_count <= 4);
 
         let is_persistent = call.is_persistent as u64;
-        let mut topic_stack_entry = if topic_count > 0 {
-            step.rw_indices[6 + call.is_persistent as usize]
-        } else {
-            // if topic_count == 0, this value will be no used anymore
-            (RwTableTag::Stack, 0usize)
-        };
+        let topic_stack_entry =
+            (topic_count > 0).then(|| step.rw_index(6 + call.is_persistent as usize));
 
         for i in 0..4 {
-            let mut topic = region.word_rlc(U256::zero());
-            if i < topic_count {
-                topic = region.word_rlc(block.rws[topic_stack_entry].stack_value());
-                self.topic_selectors[i].assign(region, offset, Value::known(F::ONE))?;
-                topic_stack_entry.1 += 1;
-            } else {
-                self.topic_selectors[i].assign(region, offset, Value::known(F::ZERO))?;
-            }
-            self.phase2_topics[i].assign(region, offset, topic)?;
+            let topic = topic_stack_entry
+                .map(|OperationRef(target, index)| {
+                    block.rws[OperationRef(target, index + i)].stack_value()
```

Comment:
I believe this could be rewritten to use `block.get_rws`. This code is reading the topic words from the stack (which can be 0,1,2,3 or 4; depending on LOG0, LOG1, LOG2, LOG3, LOG4).
And it should be easy to figure out the offset of the first one (it should be `6 + call.is_persistent as usize`).
Probably it will be easier to rewrite the loop like `for i in 0..topic_count` and leave the other values as 0

Reply:
I just saw your comments. Let me think and figure it out.

Reply:
Thank you, @ed255, for the explanation. I know how it works now. Patched at [4945a56](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1403/commits/4945a56dab4b32a906c04f3898de0337f1ebd369)